src/Runtime/DependencyInjection.php

```diff
@@ -169,11 +163,16 @@ protected function addDrushServices(Container $container, ClassLoader $loader, D
             ->invokeMethod('setProcessManager', ['process.manager']);
     }
 
-    protected function alterServicesForDrush($container, Application $application): void
+    protected function alterServicesForDrush($container, Application $application, InputInterface $input, OutputInterface $output): void
     {
         $paramInjection = $container->get('parameterInjection');
         $paramInjection->register('Symfony\Component\Console\Style\SymfonyStyle', new DrushStyleInjector());
 
+        Robo::addShared($container, 'io', DrushStyle::class)
+            ->addArguments([$input, $output]);
+        // @todo Does this alias interfere with the paramInjector above? Could we drop that feature?
```

Comment:
Parameter injection only happens when an Annotated Command is dispatched. We should probably keep this until Robo and Annotated Command are removed completely from Drush. That could happen in Drush 14, but waiting for Drush 15 would give folks time to convert old commands.

Is it necessary to put `io` in the DI container? Doing this makes it harder to re-use code, e.g. calling a command object, but capturing its output in a buffer instead of letting it go to stdout. Symfony enables this by passing $input and $output in as parameters, e.g. in the `exectue` method. Robo struggles with this due to the way the DI container is used, especially with respect to io and the logger. Parameter injection was a mechanism / attempt to go back to the Symfony model without breaking backwards compatibility. However, "optional parameters" are unusual / magical; folks don't know to look for them or use them, and as a result, the feature was seldom used.

Reply:
I think it's better to avoid putting $output, $input, $io and $logger in the container. We're a ways off from removing all of those, but I'd resist adding more in the same category.